### PR TITLE
Exposing marching cubes 

### PR DESCRIPTION
**Pull request by daviesthomas**
This PR exposes the recent refactoring of igl/marching_cubes.cpp in the python bindings. 

The marching cubes implementation in igl is no longer "copyleft" restricted and can be exposed in the API without affecting the licensing on libigl-python-bindings. 

Changes include:
- updating bindings to pull from marching cube PR commit: https://github.com/libigl/libigl/pull/1666
- exposing marching cubes in marching_cubes.cpp
- updating offset_surface and remove_duplicates as they are no longer copyleft
- test cases for marching cubes. 

There is no addition of the marching_cubes work to tutorial. If this is required please let me know where to append to existing notebooks. 

Sample usage: 

```python
import igl
import meshplot
meshplot.offline()

import numpy as np

import os
root_folder = os.getcwd()

v, f = igl.read_triangle_mesh(os.path.join(root_folder, "data", "armadillo.obj"))

#sample points on a 64x64x64 grid
n = 64
K = np.linspace( -1.0, 1.0, n)
pts = np.array([[x,y,z] for x in K for y in K for z in K])

S, _, _ = igl.signed_distance(pts, v, f, sign_type=igl.SIGNED_DISTANCE_TYPE_FAST_WINDING_NUMBER)

nV, nF = igl.marching_cubes(S, pts, n, n, n, 0.0)

meshplot.plot(nV, nF)
```

~Happy holidays! ❄️

### CODE REVIEW

**Comment by daviesthomas** [comment]
@teseoch just bumping this now that the holidays and deadline hell is over 👍  Anything I can do to move this along?

**Comment by marcelsan** [comment]
Hi @teseoch and @daviesthomas  

Do you have any update on this PR? I am working on a project that this implementation of marching cubes would be very valuable.

Thanks!

**Comment by daviesthomas** [comment]
@marcelsan whew, this completly slipped me. I'll work on this tomorrow and address @teseoch comments above. Apologies on this not already being finished!

**Comment by marcelsan** [comment]
@daviesthomas  thank you for the quick response and for working on this!

**Comment by marcelsan** [comment]
@daviesthomas do you have any update on this? Thanks!

File: src/marching_cubes.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+npe_arg(s, dense_float, dense_double)
+npe_arg(gv, dense_float, dense_double)
+npe_arg(nx, int)
+npe_arg(ny, int)
+npe_arg(nz, int)
+npe_default_arg(isovalue, double, 0.0)
+
+npe_begin_code()
+    
+    // vertices and faces of marched iso surface
```
[the remark this answers is not in the document]
Comment (daviesthomas) [reply src/marching_cubes.cpp:63]: fixed. 

Checking rows of S and GV match. 

Checking S has 1 col

Checking GV has 3 col.

File: src/marching_cubes.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    Eigen::MatrixXd SV;
+    Eigen::MatrixXi SF; 
+
+    Eigen::MatrixXd s_copy = s.template cast<double>();
+    Eigen::MatrixXd gv_copy = gv.template cast<double>();
+
+    igl::marching_cubes(s_copy, gv_copy, nx, ny, nz, isovalue, SV, SF);
+
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> svRowMajor = SV; 
+    Eigen::Matrix<int, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> sfRowMajor = SF;
```
[the remark this answers is not in the document]
Comment (daviesthomas) [reply src/marching_cubes.cpp:73]: fixed.

File: tests/test_basic.py [hunk 3]
```diff
@@ -705,6 +705,26 @@ def test_map_vertices_to_circle(self):
...
+
+        #test marching over a sphere 
+        sphereField = np.linalg.norm(pts, axis=1) - 1 
+        V,F = igl.marching_cubes(sphereField, pts, n, n, n, 0.0)
+
+        # make sure we get "something" (TODO: should we confirm this something is a sphere?)
+        self.assertNotEqual(V.shape, (0,3))
+        self.assertNotEqual(F.shape, (0,3))
+        self.assertTrue(F.flags.c_contiguous)
+        self.assertTrue(F.flags.c_contiguous)
```
[the remark this answers is not in the document]
Comment (daviesthomas) [reply tests/test_basic.py:728]: added test.

**Comment by daviesthomas** [comment]
> @daviesthomas do you have any update on this? Thanks!

Just got a chance to finish the PR up. 

If you want this functionality now, you can clone this PR locally and run 

`python setup.py install`

(I would advise doing so in a pipenv/or similar environment!)

**Comment by teseoch** [comment]
It is now merged!

**Comment by daviesthomas** [comment]
> It is now merged!

Thanks @teseoch! And sorry for that insane delay...

**Comment by marcelsan** [comment]
Thanks @teseoch  and @daviesthomas for the work! As far as I can see, there isn't an example of this function in the documentation. Should I go ahead and update it?

**Pull request merged**